api/kv.go

```diff
@@ -0,0 +1,300 @@
+type KVClient struct {
+	c         *Client
+	mountPath string
+	version   int
+}
+
+type KVSecret struct {
+	Data     map[string]interface{}
+	Metadata *VersionMetadata
+	Raw      *Secret
+}
+
+type KVMetadata struct {
+	CASRequired        bool                   `mapstructure:"cas_required"`
+	CreatedTime        time.Time              `mapstructure:"created_time"`
+	CurrentVersion     int                    `mapstructure:"current_version"`
+	CustomMetadata     map[string]interface{} `mapstructure:"custom_metadata"`
+	DeleteVersionAfter time.Duration          `mapstructure:"delete_version_after"`
+	MaxVersions        int                    `mapstructure:"max_versions"`
+	OldestVersion      int                    `mapstructure:"oldest_version"`
+	UpdatedTime        time.Time              `mapstructure:"updated_time"`
+	// Keys are stringified ints, e.g. "3"
+	Versions map[string]VersionMetadata `mapstructure:"versions"`
+}
+
+type VersionMetadata struct {
+	Version      int       `mapstructure:"version"`
+	CreatedTime  time.Time `mapstructure:"created_time"`
+	DeletionTime time.Time `mapstructure:"deletion_time"`
+	Destroyed    bool      `mapstructure:"destroyed"`
+	// There is currently no version-specific custom metadata.
+	// This field is just a copy of what's in the CustomMetadata field
+	// for the full KVMetadata of the secret.
+	CustomMetadata map[string]string `mapstructure:"custom_metadata"`
+}
+
+func (c *Client) KVv1(mountPath string) *KVClient {
+	return &KVClient{c: c, mountPath: mountPath, version: 1}
+}
+
+func (c *Client) KVv2(mountPath string) *KVClient {
+	return &KVClient{c: c, mountPath: mountPath, version: 2}
+}
+
+func (kv *KVClient) Read(ctx context.Context, secretPath string) (*KVSecret, error) {
+	pathToRead, err := kv.getFullPath(secretPath)
+	if err != nil {
+		return nil, fmt.Errorf("error assembling full path to KV secret: %v", err)
+	}
+
+	secret, err := kv.c.Logical().ReadWithContext(ctx, pathToRead)
+	if err != nil {
+		return nil, fmt.Errorf("error encountered while reading secret at %s: %v", pathToRead, err)
+	}
+	if secret == nil || secret.Data == nil {
+		return nil, fmt.Errorf("no secret found at %s", pathToRead)
+	}
+
+	return extractDataAndVersionMetadata(secret, kv.version)
+}
+
+// ReadVersion returns the data and metadata for a specific version of the
+// given secret. If that version has been deleted, the Data field on the
+// returned secret will be nil, and the Metadata field will contain the deletion time.
+func (kv *KVClient) ReadVersion(ctx context.Context, secretPath string, version int) (*KVSecret, error) {
+	pathToRead, err := kv.getFullPath(secretPath)
+	if err != nil {
+		return nil, fmt.Errorf("error assembling full path to KV secret: %v", err)
+	}
+
+	queryParams := map[string][]string{"version": {strconv.Itoa(version)}}
+	secret, err := kv.c.Logical().ReadWithDataWithContext(ctx, pathToRead, queryParams)
+	if err != nil {
+		return nil, err
+	}
+	if secret == nil || secret.Data == nil {
```

Comment:
Today, the Logical().Read() method returns `nil, nil` when a secret is not found. Many users have been confused by this behavior, not realizing they should check for nil in addition to their usual error handling. But we always kept it that way because that behavior is important for certain Vault internals (which the Logical() methods were originally written for.)

Now, however, this new KV wrapper is a chance to finally give the people what they want! A Read operation that returns an error when the requested secret doesn't exist! And that won't need to also work with those specific Vault-internal use cases! For many users, this is the only Read method they'll ever need.

However, is it worth introducing an inconsistency in secret-fetching behavior? Since other secrets engines will still need to go through Logical().Read(), which will continue to behave the way it always did. So users using both KV helpers and other secrets engines may be befuddled by this!

I'm on the side of yes it's worth it, just because I think we should aim for maximum straightforwardness with KV, our principal secrets engine, but I can be convinced!

Reply:
Actually, I'm realizing now that v1 Logical operations often return a nil secret and a nil error, and v1 is still probably a pretty common use case. So I've added an early successful return `nil, nil` for the v1 scenario. This will be true for Delete as well. 

If anyone thinks that because we'll allow `nil, nil` from v1, then I should allow nil secrets from v2 too, please let me know!

Reply:
I general, I am very much in favor of returning an error on failure. My only concern is that it will introduce inconsistent behavior in the library. My hope was to introduce it in the new generated libraries since those make no promises of backwards compatibility.

P.S. I did not see the v1 `nil, nil` return that you are referring to.

Reply:
Yeah I think you're right that discussing that as part of the breaking changes for the new generated libraries, rather than here, is probably a good call. `client.KVv2().Write()` and `client.Logical().Write()` behaving differently from each other, in the same package, might just make things more confusing. 

Okay, I'll change the code to accept `nil, nil` as a valid return value again.

(And sorry for my confusing comments here, I added that first review comment to this `ReadVersion` method arbitrarily just because it had an example of this `if secret == nil || secret.Data == nil` stuff I was talking about in it, but it made my second review comment quite confusing because this method didn't actually have any of the logic for v1 that I was talking about. My point, that v1 returns `nil, nil` after a successful Write or Delete operation, applies anywhere where there could be one of those Logical calls made against a v1 backend.)

Reply:
I spent some time looking at this with Vinny and we ended up going the opposite direction and choosing to go back to what I had with throwing errors instead of `nil, nil` after all, as when I started to write some of the other methods like `ReadMetadata`, returning `nil, nil` started to make even less sense and I had to have disclaimers in all the godoc comments... 😆 

I think this is a fine enough time to introduce a philosophy change since there's no backwards compatibility concern with these new methods after all. They're intended to break us away from the confusion of the old way of referencing KV as possible. We can think of this as a stepping-stone bridge to the new generated libraries.